odl/discr/discr_ops.py

```diff
@@ -0,0 +1,477 @@
+    dimensions `DiscreteLp.ndim`.
+
+    Calls helper function `finite_diff` for each component of the input
+    product space vector. For the adjoint of the `DiscreteDivergence`
+    operator to match the negative `DiscreteGradient` operator implicit zero
+    padding is assumed.
+    """
+
+    def __init__(self, space):
+        """Initialize a `DiscreteDivergence` operator instance.
+
+        Zero padding is assumed for the adjoint of the `DiscreteDivergence`
+        operator to match negative `DiscreteGradient` operator.
+
+        Parameters
+        ----------
+        space : `DiscreteLp`
+            The space of elements which the operator is acting on
+        """
+        if not isinstance(space, DiscreteLp):
+            raise TypeError('space {!r} is not a `DiscreteLp` '
+                            'instance.'.format(space))
+
+        self.space = space
+        super().__init__(domain=ProductSpace(space, space.grid.ndim),
+                         range=space, linear=True)
+
+    def _apply(self, x, out):
+        """Apply the divergence operator to ``x`` and store result in ``out``.
+
+        Parameters
+        ----------
+        x : ``domain`` element
+            `ProductSpaceVector` to which the divergence operator is applied to
+        out : ``range`` element
+            Output vector to which the result is written
+
+        Examples
+        --------
+        >>> from odl import Rectangle, uniform_discr_fromspace, FunctionSpace
+        >>> data = np.array([[0., 1., 2., 3., 4.],
+        ...                  [1., 2., 3., 4., 5.],
+        ...                  [2., 3., 4., 5., 6.]])
+        >>> space = FunctionSpace(Rectangle([0, 0], [3, 5]))
+        >>> disc = uniform_discr_fromspace(space, data.shape)
+        >>> div = DiscreteDivergence(disc)
+        >>> f = div.domain.element([data, data])
+        >>> div_f = div(f)
+        >>> print(div_f)
+        [[1.0, 2.0, 2.5, 3.0, 1.0],
+         [2.0, 2.0, 2.0, 2.0, -1.0],
+         [1.0, 0.0, -0.5, -1.0, -5.0]]
+        >>> g = div.range.element(data ** 2)
+        >>> adj = div.adjoint
+        >>> adj_g = adj(g)
+        >>> g.inner(div_f)
+        -119.0
+        >>> f.inner(adj_g)
+        -119.0
+        """
+
+        ndim = self.range.grid.ndim
+        dx = self.range.grid.stride
+
+        if isinstance(self.range.dspace, Rn):
+
+            arr = out.asarray()
+            tmp = np.zeros(out.shape, out.dtype, order=out.space.order)
+            for axis in range(ndim):
+                finite_diff(x[axis], out=tmp, axis=axis, dx=dx[axis],
+                            edge_order=2, zero_padding=True)
+                if axis == 0:
+                    arr[:] = tmp
+                else:
+                    arr += tmp
+
+            # self assignment in case asarray is a view, thus no overhead
+            out[:] = arr
+
+        elif isinstance(self.domain.dspace, CudaRn):
+            raise NotImplementedError('support for `CudaRn` to be '
```

Comment:
This should work, how about write a TODO: write tests, and create an issue on this.